File: ObasAnalyzerCSharp/ObasAnalyzerCSharp/ObasAnalyzerCSharp/RepositorioMetodosLaboriosos.cs
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;
using System.Collections.Immutable;
using System.Linq;
using Utilerias.ObasAnalyzerCSharp;

namespace ObasAnalyzerCSharp
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class RepositorioMetodosLaboriosos : DiagnosticAnalyzer
    {
        #region "Reglas"
        internal const string Category = "OBAS-SmellArquitectura-MVC";

        internal static readonly DiagnosticDescriptor Regla001RepositorioMetodosLaborioso = new DiagnosticDescriptor(
            "RA12001",
            "RA12-001: Repositorio con método que ejecuta varias acciones",
            "El método {0} del repositorio {1} ejecuta muchos llamados a acciones sobre la base de datos.",
            Category,
            DiagnosticSeverity.Error,
            true);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Regla001RepositorioMetodosLaborioso); } }

        public override void Initialize(AnalysisContext context)
        {
            context.EnableConcurrentExecution();
            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
            // Registra el tipo de sintaxis que se analiza para cada regla
            context.RegisterSyntaxNodeAction(AnalyzeRegla001RepositorioGordo, SyntaxKind.ClassDeclaration);
        }
        #endregion

        #region "Analizadores"

        /// <summary>
        /// Valida el cumplimiento de la regla
        /// "RA12-001: Repositorio con método que ejecuta varias acciones"
        /// </summary>
        /// <param name="context"></param>
        private void AnalyzeRegla001RepositorioGordo(SyntaxNodeAnalysisContext context)
        {
            // Obtiene la declaración de la clase
            var classDeclaration = (ClassDeclarationSyntax)context.Node;

            // Obtiene el simbolo de la clase
            var classSymbol = context.SemanticModel.GetDeclaredSymbol(classDeclaration);

            // Obtiene nombre de la clase
            var nombreClase = classDeclaration.Identifier.ValueText;

            // Revisa si el nombre de la clase contenedora inicia con la cadena "dal"
            if (nombreClase.ToLower().Contains(Constantes.nomenclaturaRepositorio))
            {
                // Obtiene la lista de metodos
                var metodos = classDeclaration.Members.OfType<BaseMethodDeclarationSyntax>();

                // Revisa cada metodo
                foreach (var metodo in metodos)
                {
                    // Obtiene el cuerpo del método
                    var cuerpoMetodo = metodo.Body;

                    // Obtiene los llamados en el método

                    var invocaciones = cuerpoMetodo.DescendantNodes().OfType<InvocationExpressionSyntax>();

                    int conteoAcciones = 0;

                    foreach (var invocacion in invocaciones)
                    {
                        if (invocacion.ToFullString().ToLower().Contains("vgo_conexion"))
                        {
                            conteoAcciones++;
                        }
                    }
                    if (conteoAcciones > Constantes.limiteAccionesBaseDatos)
                    {
                        // Intenta convertir el método a MethodDeclarationSyntax
                        var methodDeclaration = metodo as MethodDeclarationSyntax;
                        // Obtiene el nombre del método
                        string nombreMetodo;

                        // Si la conversión es exitosa obtiene el nombre
                        if (methodDeclaration != null)
                        {
                            nombreMetodo = methodDeclaration.Identifier.ValueText;

                            var diagnostic = Diagnostic.Create(Regla001RepositorioMetodosLaborioso, classDeclaration.Identifier.GetLocation(), nombreMetodo, classSymbol.Name);
                            context.ReportDiagnostic(diagnostic);
                        }
                    }
                }
            }
        }
    }

    #endregion
}

File: ObasAnalyzerVB/ObasAnalyzerVB/ObasAnalyzerVB/RepositorioMetodosLaboriosos.vb
Imports ObasAnalyzerVB.Utilerias.ObasAnalyzer

<DiagnosticAnalyzer(LanguageNames.VisualBasic)>
Public Class RepositorioMetodosLaboriosos
    Inherits DiagnosticAnalyzer

#Region "Reglas"
    Friend Const Category As String = "OBAS-SmellArquitectura"

    Friend Shared ReadOnly Regla001RepositorioMetodosLaborioso As DiagnosticDescriptor = New DiagnosticDescriptor(
            "RA12001",
            "RA12-001: Repositorio con método que ejecuta varias acciones ",
            "El método {0} del repositorio {1} ejecuta muchos llamados a acciones sobre la base de datos.",
            Category,
            DiagnosticSeverity.Error,
            True)


    Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
        Get
            Return ImmutableArray.Create(Regla001RepositorioMetodosLaborioso)
        End Get
    End Property

    Public Overrides Sub Initialize(ByVal context As AnalysisContext)
        ' Registra el tipo de sintaxis que se analiza para cada regla
        context.RegisterSyntaxNodeAction(AddressOf AnalyzeRegla001RepositorioMetodosLaborioso, SyntaxKind.ClassBlock)
    End Sub

#End Region

#Region "Analizadores"

    ''' <summary>
    ''' Valida el cumplimiento de la regla
    ''' "RA12-001: Repositorio con método que ejecuta varias acciones"
    ''' </summary>
    ''' <param name="context"></param>
    Private Sub AnalyzeRegla001RepositorioMetodosLaborioso(ByVal context As SyntaxNodeAnalysisContext)
        ' Obtiene la declaración de la clase
        Dim classDeclaration = DirectCast(context.Node, ClassBlockSyntax)

        ' Obtiene el símbolo de la clase
        Dim classSymbol = context.SemanticModel.GetDeclaredSymbol(classDeclaration)

        ' Obtiene nombre de la clase
        Dim nombreClase = classDeclaration.BlockStatement.Identifier.ValueText

        ' Revisa si el nombre de la clase contenedora inicia con la cadena "dal"
        If nombreClase.ToLower().Contains(Constantes.nomenclaturaRepositorio) Then
            ' Obtiene la lista de métodos
            Dim metodos = classDeclaration.Members.OfType(Of MethodBlockSyntax)()

            ' Revisa cada método
            For Each metodo In metodos
                ' Obtiene el cuerpo del método
                Dim cuerpoMetodo = metodo.Statements

                ' Obtiene los llamados en el método
                Dim invocaciones = cuerpoMetodo.SelectMany(Function(statement) statement.DescendantNodesAndSelf().OfType(Of InvocationExpressionSyntax)())

                Dim conteoAcciones As Integer = 0


                For Each invocacion In invocaciones
                    For Each comandoEjecucion In Constantes.commandosEjecucionSQL
                        If invocacion.ToFullString().ToLower().Contains(comandoEjecucion) Then
                            conteoAcciones += 1
                        End If
                    Next
                Next

                If conteoAcciones > Constantes.limiteAccionesBaseDatos Then
                    ' Obtiene el nombre del método
                    Dim nombreMetodo = metodo.SubOrFunctionStatement.Identifier.ValueText

                    Dim diag = Diagnostic.Create(Regla001RepositorioMetodosLaborioso, classDeclaration.BlockStatement.Identifier.GetLocation(), nombreMetodo, classSymbol.Name)
                    context.ReportDiagnostic(diag)
                End If
            Next
        End If
    End Sub

#End Region


End Class
